src/main/java/graphql/nadel/engine/ResultNodesToOverallResult.java

```diff
@@ -0,0 +1,98 @@
+package graphql.nadel.engine;
+
+import graphql.execution.ExecutionStepInfo;
+import graphql.execution.nextgen.FetchedValueAnalysis;
+import graphql.execution.nextgen.result.ExecutionResultNode;
+import graphql.execution.nextgen.result.LeafExecutionResultNode;
+import graphql.execution.nextgen.result.ListExecutionResultNode;
+import graphql.execution.nextgen.result.ObjectExecutionResultNode;
+import graphql.execution.nextgen.result.RootExecutionResultNode;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLOutputType;
+import graphql.schema.GraphQLSchema;
+import graphql.util.TraversalControl;
+import graphql.util.TraverserContext;
+import graphql.util.TraverserVisitorStub;
+import graphql.util.TreeTransformerUtil;
+
+import static graphql.Assert.assertShouldNeverHappen;
+
+public class ResultNodesToOverallResult {
+
+
+    public RootExecutionResultNode convert(RootExecutionResultNode resultNode, GraphQLSchema overallSchema) {
```

Comment:
I am not sure about returning RootExecutionResultNode here, how would it work with hydration? Since the input of the service query is a list of fields (e.g. toplevel fields for a given service that were batched together or hydration fields from previous query that will be top level fields in the service query) we should return here a Map of <FieldPathInTheOverallQuery, ExecutionResultNode>. You cannot just call merge trees on the result of this function as fields resolved by delegate may be on the different levels in the overall result.

E.g. say we have overall Schema like so: `Page { author: Account, owner: Account,  id: ID! }` and query that asks about it.
Then we will generate 1st delegate query like so: `Page { id authorId ownerId}` after translation I would expect result like that:
/Page = ObjectResultNode(Page, UnresolvedHydration(author), UnresolvedHydration(owner))

Then you may merge it to overall schema using paths (e.g. if there were other top level query fields they will be there) and then collect unresolved nodes. which will lead to to second delegate query:
`author: user(id: $authorId) { ... } owner: user(id: ownerId)` and now the result of mapping should be like so:
/Page/author = ObjectResultNode(allFields from author result node)
/Page/owner = ObjectResultNode(allFields from owner result node)
Merge with overall result collect unresolved and do it again until there are no unresolved nodes

Reply:
added comment